dashboard/test/controllers/regional_partners_controller_test.rb

```diff
@@ -85,4 +88,78 @@ def self.test_workshop_admin_only(method, action, params = nil)
     assert @regional_partner.program_managers.exists?(teacher.id)
     assert teacher.program_manager?
   end
+
+  test 'add mapping updates regional partner mapping' do
+    sign_in @workshop_admin
+    post :add_mapping, params: {id: @regional_partner.id, region: '12345'}
+    assert @regional_partner.mappings.length == 1
+  end
+
+  test 'replace mappings fails on non-csv file' do
+    sign_in @workshop_admin
+    mapping = fixture_file_upload('regional_partners.tsv', 'text/tsv')
+    post :replace_mappings, params: {id: @regional_partner.id, regions: mapping}
+    assert_nil flash[:notice]
+    assert_equal(
+      'Replace mappings failed. File is not a CSV.',
+      flash[:alert]
+    )
+    assert @regional_partner.reload.mappings.empty?
+  end
+
+  test 'replace mappings on invalid mapping fails and does not delete old mapping' do
+    regional_partner_with_mapping = create :regional_partner_with_mappings
+    sign_in @workshop_admin
+    mapping = fixture_file_upload('regional_partner_mappings_invalid.csv', 'text/csv')
+    post :replace_mappings, params: {id: regional_partner_with_mapping.id, regions: mapping}
+    assert_nil flash[:notice]
+    assert_equal(
+      '<b>Replace mappings failed with 2 error(s):</b><br>ABC: Invalid region<br>123456: Invalid region',
+      flash[:upload_error]
+    )
+    refute regional_partner_with_mapping.mappings.empty?
+    assert_equal '98143', regional_partner_with_mapping.mappings.first.zip_code
+  end
+
+  test 'replace mappings fails on invalid non-unique mapping' do
```

Comment:
nit: "non-unique zipcode mapping" to distinguish with the test for "non-unique state mapping".